### PR TITLE
Feat/15 gen random

### LINKED ISSUE #15: generate random sequence from grammar
As was discussed in #7 , generating a random sequence from a BNF grammar is a good feature. The [erratic.js](https://github.com/dhconnelly/erratic) serves as a good example of behavior to implement. My only question is:

- should the generated sequence be returned as a `Vec<Terminals>` or as a `String`?

My assumptions for this feature are:

- pseudo random is sufficient
- allowing users to tweak the probabilities per node is out of scope

### PR DESCRIPTION
**Pull request by snewt**
This PR represents the addition of a feature to generate random sentences from an existing grammar object with a `generate` trait. The `generate` trait returns a result in order to handle problematic grammars and report what happened. 

Documentation was updated as well as the test suite. 

Closes #15

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,108 +7,162 @@
...
-## Example
+Once the `Grammar` object is populated, to generate a random sentence from it
+call the object's generate function. `grammar.generate()`. For the above grammar
+you could expect something like `TGGC` or `AG`. 
+
+If the generate function can't find a production for a nonterminal it tries
+to evaluate it will print the identifer as a nonterminal, i.e. `<identifier>`.
+
+The generate function will return an error if it detects an infinite loop caused
+by a production such as `<PATTERN> := <PATTERN>`.
```
Comment (CrockAgile) [inline README.md]: `:=` or `::=` ?

File: README.md [hunk 2]
```diff
@@ -7,108 +7,162 @@
...
-               <zip-part> ::= <town-name> \",\" <state-code> <ZIP-code> <EOL>;
+        <opt-suffix-part> ::= \"Sr.\" | \"Jr.\" | <roman-numeral> | \"\"
+            <opt-apt-num> ::= <apt-num> | \"\";"
 
-        <opt-suffix-part> ::= \"Sr.\" | \"Jr.\" | <roman-numeral> | \"\";
-            <opt-apt-num> ::= <apt-num> | \"\";";
+    let grammar = Grammar::from_str(input);
+    match grammar {
+        Ok(g) => println!("{:#?}", g),
+        Err(e) => println!("Failed to make grammar from String: {:?}", e),
```
Comment (CrockAgile) [inline README.md]: Do bnf errors impl display? If not, they should. If they already do, then just no debug `:?` is necessary. Is the debug easier to read than the impl display?

File: src/error.rs [hunk 3]
```diff
@@ -58,3 +60,72 @@ impl From<Needed> for Error {
         Error::ParseIncomplete(string)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    // use super::*;
```
Comment (CrockAgile) [inline src/error.rs]: Is this a clarifying comment or a leftover?

File: src/grammar.rs [hunk 4]
```diff
@@ -48,19 +52,111 @@ impl Grammar {
...
-    /// Get iterator of the `Grammar`'s `Productions`s
+    /// Get iterator of the `Grammar`'s `Production`s
     pub fn productions_iter(&self) -> Iter {
         Iter {
             iterator: self.productions.iter(),
         }
     }
 
-    /// Get mutable iterator of the `Grammar`'s `Productions`s
+    /// Get mutable iterator of the `Grammar`'s `Production`s
```
Comment (CrockAgile) [inline src/grammar.rs]: Good catch!

File: src/grammar.rs [hunk 5]
```diff
@@ -48,19 +52,111 @@ impl Grammar {
...
+    fn traverse(&self, ident: &String) -> Result<String, Error> {
+        let stack_red_zone: usize = 32 * 1024; // 32KB
+        if stacker::remaining_stack() <= stack_red_zone {
+            return Err(Error::GenerateError(
+                String::from("Infinite loop detected!"),
+            ));
+        }
+
+        let nonterm = Term::Nonterminal(ident.clone());
+        let mut result = String::new();
```
Comment (CrockAgile) [inline src/grammar.rs]: Definitely a small point, but what do you think about moving `result` down to where it is used? As I was reading the code I had assigned some mental effort to tracking its mutations and was surprised to find none until the very end.

File: src/grammar.rs [hunk 6]
```diff
@@ -48,19 +52,111 @@ impl Grammar {
...
+    ///     let input =
+    ///         "<dna> ::= <base> | <base> <dna>
+    ///         <base> ::= \"A\" | \"C\" | \"G\" | \"T\"";
+    ///     let grammar = Grammar::from_str(input).unwrap();
+    ///     let sentence = grammar.generate();
+    ///     match sentence {
+    ///         Ok(s) => println!("random sentence: {}", s),
+    ///         Err(e) => println!("something went wrong: {}!", e)
+    ///     }
+    /// }
```
Comment (CrockAgile) [inline src/grammar.rs]: This doc string example made me realize that for `Expression::remove_term` or w/e the example hides the `extern`/`use`/`main` etc. If we want that context in these small examples, then that one should be updated to match this.

File: src/grammar.rs [hunk 7]
```diff
@@ -48,19 +52,111 @@ impl Grammar {
...
+    /// ```
```
Comment (CrockAgile) [inline src/grammar.rs]: Maybe overkill, but can you think of any hidden `assert`s at the end of this example that would help prevent it from going stale in the future? If not, then no worries, not a must.

File: src/grammar.rs [hunk 8]
```diff
@@ -48,19 +52,111 @@ impl Grammar {
...
+    pub fn generate(&self) -> Result<String, Error> {
+        let start_rule: String;
+        let lhs = self.productions_iter().nth(0);
```
Comment (CrockAgile) [inline src/grammar.rs]: This variable is the first production, not the lhs as the name suggests. And some of the name confusion continues in the match below, referring to the production as a term.

File: README.md (hunk 1 above)
Comment (snewt) [reply README.md]: :+1:

File: README.md (hunk 2 above)
Comment (snewt) [reply README.md]: Ah no they shouldn't need debug, it's just a bad habit I picked up to default to always use `:?`

File: src/grammar.rs (hunk 7 above)
Comment (snewt) [reply src/grammar.rs]: can you give an example? assert what?

File: src/grammar.rs (hunk 6 above)
Comment (snewt) [reply src/grammar.rs]: yeah I kind of like it, I'll update it

File: src/grammar.rs (hunk 7 above)
Comment (CrockAgile) [reply src/grammar.rs]: `///# assert(sentence.is_ok())` or something would be enough for me. Wouldn't show up in the docs, but would be another canary if somehow the example gets messed with in the future and no longer parses or generates successfully.

**Pull request merged**